core/src/main/scala/org/apache/spark/util/Ordering.scala

```diff
@@ -0,0 +1,42 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.util
+
+private[spark] class HashOrdering[A] extends Ordering[A] {
+  override def compare(x: A, y: A): Int = {
+    val h1 = if (x == null) 0 else x.hashCode()
+    val h2 = if (y == null) 0 else y.hashCode()
+    if (h1 < h2) -1 else if (h1 == h2) 0 else 1
+  }
+}
+
+private[spark] class NoOrdering[A] extends Ordering[A] {
+  override def compare(x: A, y: A): Int = 0
+}
+
+private[spark] class KeyValueOrdering[A, B](
+  ordering1: Option[Ordering[A]], ordering2: Option[Ordering[B]]
+) extends Ordering[Product2[A, B]] {
+  private val ord1 = ordering1.getOrElse(new HashOrdering[A])
+  private val ord2 = ordering2.getOrElse(new NoOrdering[B])
+
+  override def compare(x: Product2[A, B], y: Product2[A, B]): Int = {
+    val c1 = ord1.compare(x._1, y._1)
+    if (c1 != 0) c1 else ord2.compare(x._2, y._2)
```

Comment:
i see 2 options:
1) do something similar to what happens in ExternalSorter.mergeWithAggregation where in groupByKeyAndSortValues i am aware of the fact that i might be processing multiple keys (with same hashCode) at once and check for key equality. this increases memory requirements
(all values for all keys with same hashCode have to fit in memory as opposed to all values for a single key).
2) require an ordering for K which can be used as a tie breaker when the hashCodes of the keys are the same, so that i have a total ordering for K.

thoughts?

i will add a unit test where i have multiple keys with the same hashCode.